packages/fuels-programs/src/lib.rs

```diff
@@ -21,3 +32,323 @@ impl Configurables {
         }
     }
 }
+
+pub struct ReceiptParser {
+    receipts: Vec<Receipt>,
+}
+
+impl ReceiptParser {
+    pub fn new(receipts: &[Receipt]) -> Self {
+        let relevant_receipts: Vec<Receipt> = receipts
+            .iter()
+            .filter(|receipt| {
+                matches!(receipt, Receipt::ReturnData { .. } | Receipt::Return { .. })
+            })
+            .cloned()
+            .collect();
+
+        Self {
+            receipts: relevant_receipts,
+        }
+    }
+
+    /// Based on receipts returned by a script transaction, the contract ID (in the case of a contract call),
+    /// and the output param, parse the values and return them as Token.
+    pub fn try_parse_output(
+        &mut self,
+        contract_id: Option<&Bech32ContractId>,
+        output_param: &ParamType,
+    ) -> Result<Token> {
+        let null_contract_id = ContractId::new([0u8; 32]);
+        let contract_id: ContractId = match contract_id {
+            Some(contract_id) => contract_id.into(),
+            // During a script execution, the script's contract id is the **null** contract id
+            None => null_contract_id,
+        };
+
+        let encoded_value = match output_param.get_return_location() {
+            ReturnLocation::ReturnData if output_param.is_vm_heap_type() => {
+                self.extract_return_data_heap(contract_id)
+            }
+            ReturnLocation::ReturnData => self.extract_return_data(contract_id),
+            ReturnLocation::Return => self.extract_return(contract_id),
+        };
+
+        if let Some(value) = encoded_value {
+            //self.receipts.remove(index);
+            let decoded_value = ABIDecoder::decode_single(output_param, &value)?;
+            Ok(decoded_value)
+        } else {
+            Err(Self::parsing_error(output_param))
+        }
+    }
+
+    fn parsing_error(output_param: &ParamType) -> Error {
```

Comment:
Maybe calling it `missing_receipts_error` since it is private and only describes that case?